File: jevlet/assistant/planner.py
"""Command -> plan with calibrated confidence, risk, and a gate; then execution.

Phase 1 asks the skill and risk questions together; phase 2 asks only the chosen skill's
argument questions. Both are parallel packed passes over the same state (the command plus the
active window). A multi-step command ("open Spotify then play") is split on "then"; later steps
are re-planned against the live screen right before they run.
"""

from __future__ import annotations

import re
import time
from dataclasses import dataclass, field

from jevlet.benchmarks import RISK_QUESTION, daily_state
from jevlet.system_one import ChoiceQuestion, NoulQuestion, SystemOne

from . import actions
from .environment import (
    KNOWN_FOLDERS,
    SETTINGS_PAGES,
    SHORTCUTS,
    WEBSITES,
    App,
    Window,
    installed_apps,
    open_windows,
)
from .skills import (
    NOT_APPLICABLE,
    SKILL_BY_NAME,
    SKILL_QUESTION,
    SKILLS,
    SLOTS,
    Environment,
    Skill,
    slot_options,
    window_label,
)
from .text import parse_duration

SAFE_TO_RUN = 0.9  # Jev's bar for destructive operations: P(safe) must reach 0.9
CONFIDENT = 0.75  # skill and argument confidence needed to run on a single Enter


@dataclass(slots=True)
class Context:
    """What was on screen when the palette opened (the palette itself is excluded)."""

    current: Window | None
    windows: list[Window]
    apps: list[App]

    @classmethod
    def capture(cls, exclude_handles: tuple[int, ...] = ()) -> Context:
        windows = open_windows(exclude_handles)
        return cls(windows[0] if windows else None, windows, installed_apps())

    def environment(self) -> Environment:
        return Environment(
            apps=[app.name for app in self.apps],
            windows=[window_label(w.process, w.title) for w in self.windows],
            current_window=window_label(self.current.process, self.current.title)
            if self.current
            else "",
        )


@dataclass(slots=True)
class Plan:
    command: str
    skill: Skill
    confidence: float
    risk: float
    args: dict[str, str] = field(default_factory=dict)
    arg_confidence: dict[str, float] = field(default_factory=dict)
    alternatives: list[tuple[Skill, float]] = field(default_factory=list)
    duration: int | None = None
    control: object | None = None
    latency_ms: float = 0.0
    missing: str = ""

    @property
    def title(self) -> str:
        values = {slot: _display(slot, value) for slot, value in self.args.items()}
        values["duration"] = _format_duration(self.duration)
        if self.control is not None:
            values["control"] = f"“{self.control.name or self.control.automation_id}”"
        try:
            return self.skill.title.format(**values)
        except KeyError:
            return self.skill.name

    @property
    def gate(self) -> str:
        """run: one Enter executes. confirm: a second Enter is required. clarify: cannot run."""
        if self.skill.key == "clarify" or self.missing:
            return "clarify"
        weakest = min([self.confidence, *self.arg_confidence.values()])
        if 1.0 - self.risk >= SAFE_TO_RUN and weakest >= CONFIDENT:
            return "run"
        return "confirm"


def _display(slot: str, value: str) -> str:
    """Human wording for the UI; the model's options keep the process name for precision."""
    if slot != "window":
        return value
    if value.startswith("Current window"):
        return "this window"
    return value.split(": ", 1)[1] if ": " in value else value


def _format_duration(seconds: int | None) -> str:
    if not seconds:
        return "?"
    minutes, second = divmod(seconds, 60)
    hours, minute = divmod(minutes, 60)
    parts = [
        f"{hours} h" if hours else "",
        f"{minute} min" if minute else "",
        f"{second} s" if second else "",
    ]
    return " ".join(part for part in parts if part)


def split_steps(command: str) -> list[str]:
    steps = re.split(
        r"\s*(?:,\s*)?\b(?:and then|then|after that)\b\s*", command.strip(), flags=re.I
    )
    return [step.strip(" ,.") for step in steps if step.strip(" ,.")]


class Planner:
    def __init__(self, engine: SystemOne) -> None:
        self.engine = engine
        # Many argument options can exceed training-time packing limits; positions still fit.
        config = getattr(self.engine.collator, "config", None)
        if config is not None:
            config.max_packed_len = max(config.max_packed_len, 2048)

    def _state(self, command: str, context: Context) -> str:
        current = context.current
        return daily_state(command, window_label(current.process, current.title) if current else "")

    def plan(self, command: str, context: Context) -> Plan:
        started = time.perf_counter()
        state = self._state(command, context)
        first = self.engine.evaluate(
            state,
            {
                "skill": ChoiceQuestion(SKILL_QUESTION, {s.name: s.description for s in SKILLS}),
                "risk": NoulQuestion(RISK_QUESTION),
            },
        )
        choice = first["skill"]
        ranked = sorted(
            zip(choice.options, choice.probabilities, strict=True), key=lambda pair: -pair[1]
        )
        skill = SKILL_BY_NAME[ranked[0][0]]
        plan = self.plan_for(skill, command, context, ranked[0][1], first["risk"].probability_true)
        plan.alternatives = [(SKILL_BY_NAME[name], p) for name, p in ranked[1:4]]
        plan.latency_ms = (time.perf_counter() - started) * 1000
        return plan

    def plan_for(
        self, skill: Skill, command: str, context: Context, confidence: float, risk: float
    ) -> Plan:
        """Fill one skill's arguments (used for the top skill and for chosen alternatives)."""
        plan = Plan(command, skill, confidence, max(risk, skill.risk_floor))
        env = context.environment()
        questions = {}
        for slot in skill.slots:
            options = slot_options(slot, command, env)
            if len(options) > 1:
                questions[slot] = ChoiceQuestion(SLOTS[slot].question, options)
            else:
                plan.missing = slot
        if questions:
            answers = self.engine.evaluate(self._state(command, context), questions)
            for slot, answer in answers.items():
                plan.args[slot] = answer.selected
                plan.arg_confidence[slot] = answer.confidence
                if answer.selected == NOT_APPLICABLE:
                    plan.missing = slot
        if skill.key == "timer":
            plan.duration = parse_duration(command)
            if not plan.duration:
                plan.missing = "duration"
        if skill.needs_screen:
            self._ground(plan, context)
        return plan

    def _ground(self, plan: Plan, context: Context) -> None:
        from jevlet.desktop.uia import FastUIA
        from jevlet.grounding import NONE_OF_THESE, ground

        if context.current is None:
            plan.missing = "control"
            return
        controls = FastUIA().read(context.current.handle)
        target, choice, risk = ground(self.engine, plan.command, context.current.title, controls)
        plan.control = target
        plan.arg_confidence["control"] = choice.confidence
        plan.risk = max(plan.risk, risk)
        if target is None or choice.selected == NONE_OF_THESE:
            plan.missing = "control"


def _window_for(label: str, context: Context) -> Window | None:
    if label.startswith("Current window") and context.current:
        return context.current
    for window in context.windows:
        if window_label(window.process, window.title) == label:
            return window
    return None


def execute(plan: Plan, context: Context) -> actions.Outcome:
    key, args = plan.skill.key, plan.args
    previous = context.current.handle if context.current else None
    if key == "open_app":
        app = next(app for app in context.apps if app.name == args["app"])
        return actions.launch_app(app.app_id, app.name)
    if key in {"switch_window", "close_window", "minimize_window", "maximize_window"}:
        window = _window_for(args["window"], context)
        if window is None:
            return actions.Outcome(False, "That window is no longer open")
        if key == "switch_window":
            return actions.switch_to(window.handle, window.title, previous)
        return actions.window_state(window.handle, window.title, key.split("_")[0])
    if key == "media":
        return actions.media(args["media"])
    if key == "volume":
        return actions.volume(args["volume"])
    if key == "settings":
        return actions.open_target(SETTINGS_PAGES[args["setting"]], args["setting"])
    if key == "theme":
        return actions.theme(args["theme"])
    if key == "search":
        return actions.web_search(args["text"])
    if key == "website":
        return actions.open_target(WEBSITES[args["website"]], args["website"])
    if key == "folder":
        return actions.open_target(KNOWN_FOLDERS[args["folder"]], args["folder"])
    if key == "type":
        return actions.type_text(previous, args["text"])
    if key == "shortcut":
        return actions.shortcut(args["shortcut"], SHORTCUTS[args["shortcut"]], previous)
    if key == "click":
        return actions.click_control(previous, plan.control)
    if key == "screenshot":
        return actions.screenshot()
    if key == "lock":
        return actions.lock()
    if key == "ask_ai":
        return actions.ask_ai(args["service"], args["text"])
    if key == "timer":
        return actions.Outcome(True, f"Timer set for {_format_duration(plan.duration)}")
    return actions.Outcome(False, "I need more detail to do that")

File: app/src/Jevlet.Core/Planner.cs
using System.Diagnostics;
using System.Text.RegularExpressions;

namespace Jevlet.Core;

/// <summary>What the model decided for one command, with the gate that governs running it.</summary>
public sealed class Plan(string command, Skill skill, double confidence, double risk)
{
    public const double SafeToRun = 0.9; // Jev's bar for destructive operations: P(safe) >= 0.9
    public const double Confident = 0.75; // skill and argument confidence needed for one Enter

    public string Command { get; } = command;
    public Skill Skill { get; } = skill;
    public double Confidence { get; } = confidence;
    public double Risk { get; set; } = Math.Max(risk, skill.RiskFloor);
    public Dictionary<string, string> Arguments { get; } = new(StringComparer.Ordinal);
    public Dictionary<string, double> ArgumentConfidence { get; } = new(StringComparer.Ordinal);
    public List<(Skill Skill, double Probability)> Alternatives { get; } = [];
    public int? DurationSeconds { get; set; }
    public TimeMatch? When { get; set; }
    public string Missing { get; set; } = "";
    public double LatencyMs { get; set; }
    public DateTime Now { get; init; } = DateTime.Now;

    /// <summary>run: one Enter executes. confirm: a second Enter is required. clarify: cannot run.</summary>
    public string Gate
    {
        get
        {
            if (Skill.Key == "clarify" || Missing.Length > 0)
            {
                return "clarify";
            }
            var weakest = ArgumentConfidence.Values.Append(Confidence).Min();
            return 1.0 - Risk >= SafeToRun && weakest >= Confident ? "run" : "confirm";
        }
    }

    /// <summary>The skill's title with its arguments filled in, or the skill name.</summary>
    public string Title
    {
        get
        {
            var title = Skill.Title;
            foreach (var (slot, value) in Arguments)
            {
                title = title.Replace($"{{{slot}}}", Display(slot, value), StringComparison.Ordinal);
            }
            var when = DurationSeconds is { } seconds ? FormatDuration(seconds) : When?.Describe(Now) ?? "";
            if (when.Length > 0)
            {
                title = title.Replace("{when}", when, StringComparison.Ordinal);
            }
            title = title.Replace(" {when}", "", StringComparison.Ordinal);
            return title.Contains('{', StringComparison.Ordinal) ? Skill.Name : title;
        }
    }

    /// <summary>Human wording; the model's options keep the process name for precision.</summary>
    public static string Display(string slot, string value)
    {
        if (slot != "window")
        {
            return value;
        }
        if (value.StartsWith("Current window", StringComparison.Ordinal))
        {
            return "this window";
        }
        var split = value.IndexOf(": ", StringComparison.Ordinal);
        return split >= 0 ? value[(split + 2)..] : value;
    }

    public static string FormatDuration(int seconds)
    {
        var hours = seconds / 3600;
        var minutes = seconds % 3600 / 60;
        var rest = seconds % 60;
        var parts = new List<string>();
        if (hours > 0) parts.Add($"{hours} h");
        if (minutes > 0) parts.Add($"{minutes} min");
        if (rest > 0) parts.Add($"{rest} s");
        return parts.Count > 0 ? string.Join(' ', parts) : "?";
    }
}

/// <summary>
/// Command to plan (port of jevlet/assistant/planner.py). Phase 1 asks the skill and risk
/// questions together; phase 2 asks only the chosen skill's argument questions, with options
/// from the live snapshot. Times and durations are parsed, never predicted.
/// </summary>
public sealed class Planner(IDecisionModel engine)
{
    private static readonly Regex Steps = new(@"\s*(?:,\s*)?\b(?:and then|then|after that)\b\s*",
        RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);

    public IDecisionModel Engine { get; } = engine;

    /// <summary>"open spotify then play the next song" becomes two steps.</summary>
    public static IReadOnlyList<string> SplitSteps(string command) =>
        Steps.Split(command.Trim()).Select(s => s.Trim(' ', ',', '.')).Where(s => s.Length > 0).ToList();

    public Plan Plan(string command, Snapshot snapshot, DateTime now)
    {
        var watch = Stopwatch.StartNew();
        var catalogue = Catalogue.Shared;
        var state = catalogue.State(command, snapshot.CurrentWindow);
        var answers = Engine.Evaluate(state,
        [
            new Question(catalogue.SkillQuestion, catalogue.Skills.Select(s => s.Name).ToList()),
            new Question(catalogue.RiskQuestion, ["True", "False"], "noul"),
        ]);
        var skills = answers[0];
        var ranked = skills.Options.Zip(skills.Probabilities).OrderByDescending(p => p.Second).ToList();
        var plan = PlanFor(catalogue.ByName[ranked[0].First], command, snapshot, ranked[0].Second, answers[1]["True"], now);
        plan.Alternatives.AddRange(ranked.Skip(1).Take(3).Select(p => (catalogue.ByName[p.First], p.Second)));
        plan.LatencyMs = watch.Elapsed.TotalMilliseconds;
        return plan;
    }

    /// <summary>Fill one skill's arguments (used for the top skill and for chosen alternatives).</summary>
    public Plan PlanFor(Skill skill, string command, Snapshot snapshot, double confidence, double risk, DateTime now)
    {
        var catalogue = Catalogue.Shared;
        var plan = new Plan(command, skill, confidence, risk) { Now = now };
        var slots = skill.Slots.Concat(skill.OptionalSlots).ToList();
        var questions = new List<(string Slot, Question Question)>();
        foreach (var slot in slots)
        {
            var options = SlotOptions.For(slot, command, snapshot);
            if (options.Count > 1)
            {
                questions.Add((slot, new Question(catalogue.SlotQuestions[slot], options)));
            }
            else if (skill.Slots.Contains(slot))
            {
                plan.Missing = slot;
            }
        }
        if (questions.Count > 0)
        {
            var answers = Engine.Evaluate(catalogue.State(command, snapshot.CurrentWindow), questions.Select(q => q.Question).ToList());
            for (var index = 0; index < questions.Count; index++)
            {
                var (slot, _) = questions[index];
                var answer = answers[index];
                if (answer.Selected == catalogue.NotApplicable)
                {
                    if (skill.Slots.Contains(slot))
                    {
                        plan.Missing = slot;
                    }
                    continue; // an optional argument the command did not give
                }
                plan.Arguments[slot] = answer.Selected;
                plan.ArgumentConfidence[slot] = answer.Confidence;
            }
        }
        switch (skill.Time)
        {
            case "duration":
                plan.DurationSeconds = TextRules.ParseDuration(command);
                if (plan.DurationSeconds is null)
                {
                    plan.Missing = "duration";
                }
                break;
            case "required":
                plan.When = TimeParser.Parse(command, now);
                if (plan.When is null)
                {
                    plan.Missing = "time";
                }
                break;
            case "optional":
                plan.When = TimeParser.Parse(command, now);
                break;
        }
        return plan;
    }
}
